types/coin.go

```diff
@@ -246,63 +322,123 @@ func (coins Coins) safePlus(coinsB Coins) Coins {
 	}
 }
 
-// Minus subtracts a set of coins from another.
-//
-// e.g.
-// {2A, 3B} - {A} = {A, 3B}
-// {2A} - {0B} = {2A}
-// {A, B} - {A} = {B}
-//
-// CONTRACT: Minus will never return Coins where one Coin has a non-positive
-// amount. In otherwords, IsValid will always return true.
-func (coins Coins) Minus(coinsB Coins) Coins {
-	diff, hasNeg := coins.SafeMinus(coinsB)
-	if hasNeg {
-		panic("negative coin amount")
+// ContainsDenomsOf returns true if coinsB' denom set
+// is subset of the receiver's denoms.
+func (coins Coins) ContainsDenomsOf(coinsB Coins) bool {
+	// more denoms in B than in receiver
+	if len(coinsB) > len(coins) {
+		return false
+	}
+
+	for _, coinB := range coinsB {
+		if coins.AmountOf(coinB.Denom).IsZero() {
+			return false
+		}
 	}
 
-	return diff
+	return true
 }
 
-// SafeMinus performs the same arithmetic as Minus but returns a boolean if any
-// negative coin amount was returned.
-func (coins Coins) SafeMinus(coinsB Coins) (Coins, bool) {
-	diff := coins.safePlus(coinsB.negative())
-	return diff, diff.IsAnyNegative()
+// SameDenomsOf returns true if both Coin sets have the
+// very same denoms.
+func (coins Coins) SameDenomsOf(coinsB Coins) bool {
+	return (len(coins) == len(coinsB)) && coins.ContainsDenomsOf(coinsB)
 }
 
 // IsAllGT returns true if for every denom in coins, the denom is present at a
 // greater amount in coinsB.
 func (coins Coins) IsAllGT(coinsB Coins) bool {
-	diff, _ := coins.SafeMinus(coinsB)
-	if len(diff) == 0 {
+	if len(coins) == 0 {
 		return false
 	}
 
-	return diff.IsAllPositive()
+	if len(coinsB) == 0 {
+		return true
+	}
+
+	if !coins.ContainsDenomsOf(coinsB) {
+		return false
+	}
+
+	for _, coinB := range coinsB {
+		amountA, amountB := coins.AmountOf(coinB.Denom), coinB.Amount
+		if !amountA.GT(amountB) {
+			return false
+		}
+	}
+
+	return true
 }
 
-// IsAllGTE returns true iff for every denom in coins, the denom is present at
-// an equal or greater amount in coinsB.
+// IsAllGT returns false if for any denom in coins,
+// the denom is present in coinsB at a greater or
+// equal amount
 func (coins Coins) IsAllGTE(coinsB Coins) bool {
-	diff, _ := coins.SafeMinus(coinsB)
-	if len(diff) == 0 {
+	if len(coinsB) == 0 {
 		return true
 	}
 
-	return !diff.IsAnyNegative()
+	if len(coins) == 0 {
+		return false
+	}
+
+	if !coinsB.Difference(coins).IsZero() {
+		return false
+	}
+
+	for _, coin := range coins {
+		if coin.Amount.LT(coinsB.AmountOf(coin.Denom)) {
+			return false
+		}
+	}
+
+	return true
 }
 
-// IsAllLT returns True iff for every denom in coins, the denom is present at
-// a smaller amount in coinsB.
+// IsAllLT returns true if all denoms in coins
+// are less than their counterparts in coinsB.
 func (coins Coins) IsAllLT(coinsB Coins) bool {
```

Comment:
why not just use IsAllGT?

Reply:
I agree, `IsAllLT` can actually be removed - it's unused.